### PR TITLE
Fix for cloud fetch

### PR DESCRIPTION
**Pull request by andrefurlan-db**
- Throw when failed to download file

- Retry properly while downloading file
- Add a bunch of debug logs
- Prevent thread issues

TODO: http connection pools for cloud storage, proxies, etc.

Also backported to version 2

### CODE REVIEW

File: src/databricks/sql/cloudfetch/download_manager.py [hunk 1]
```diff
@@ -97,9 +98,11 @@ def get_next_downloaded_file(
             self.download_handlers.pop(idx)
             # Return True upon successful download to continue loop and not force a retry
             return result
-        # Download was not successful for next download item, force a retry
+        # Download was not successful for next download item. Fail
         self._shutdown_manager()
-        return None
+        raise ResultSetDownloadError(
```
[the remark this answers is not in the document]
Comment (benc-db) [reply src/databricks/sql/cloudfetch/download_manager.py:103]: Or is it just handled by raising the exception?

[the remark this answers is not in the document]
Comment (andrefurlan-db) [reply src/databricks/sql/cloudfetch/download_manager.py:103]: the retry is done outside this function, closer to the actual http request

File: src/databricks/sql/cloudfetch/downloader.py [hunk 2]
```diff
@@ -171,3 +202,40 @@ def decompress_data(compressed_data: bytes) -> bytes:
                 uncompressed_data += data
                 start += num_bytes
         return uncompressed_data
+
+
+def http_get_with_retry(url, max_retries=5, backoff_factor=2, download_timeout=60):
```
Comment (benc-db) [inline src/databricks/sql/cloudfetch/downloader.py:207]: why are we implementing retry behavior here rather than using a Retry passed to the session?

**Review by benc-db — APPROVED** [review]
Approve, but consider if we can implement with urllib3 Retry.

File: src/databricks/sql/cloudfetch/downloader.py (hunk 2 above)
Comment (andrefurlan-db) [reply src/databricks/sql/cloudfetch/downloader.py:207]: agreed. It is in the TODO to also have connection pools

**Pull request closed without merging**